// megu-randbooru/index.js
const superagent = require('superagent');
const parseString = require('xml2js').parseString;

function testArrayNsfw(result) {
    var valid = true;
    try {
        if (typeof (result) == 'undefined') valid = false;
        if (typeof (result.posts) == 'undefined') valid = false;
        if (typeof (result.posts.post) == 'undefined') valid = false;
        if (typeof (result.posts.post[0]) == 'undefined') valid = false;
        if (typeof (result.posts.post[0].file_url) == 'undefined') valid = false;
        if (typeof (result.posts.post[0].id) == 'undefined') valid = false;
    } catch(ex) {
        valid = false;
    }
    
    return valid;
}

function testArraySfw(result) {
    var valid = true;
    try {
        if (typeof (result) == 'undefined') valid = false;
        if (typeof (result.posts) == 'undefined') valid = false;
        if (typeof (result.posts.post) == 'undefined') valid = false;
        if (typeof (result.posts.post[0]) == 'undefined') valid = false;
        if (typeof (result.posts.post[0].$.file_url) == 'undefined') valid = false;
        if (typeof (result.posts.post[0].$.id) == 'undefined') valid = false;
    } catch(ex) {
        valid = false;
    }
    
    return valid;
}

class BooruGrabber {
    /**
     * 
     * @param {String} type "sfw" or "nsfw"
     */
    constructor(type) {
        if (type == "sfw") {
            this.safe = true;
            this.url = `https://safebooru.org/index.php`;
        } else if (type == "nsfw") {
            this.safe = false;
            this.url = `http://gelbooru.com/index.php`;
        } else throw new Error("You must specify either 'sfw' or 'nsfw' as a parameter")
    }

    randomImage(tags) {
        return new Promise((resolve, reject) => {
            this._getImages(1, 0, tags).then(data => {
                let clean = data.text.replace("\ufeff", "");
                parseString(clean, (err, result) => {
                    if (err) reject(err);
                    let randomPid;
                    if(result.posts.$.count >= 20000){
                        randomPid = Math.floor(Math.random() * 20000);
                    } else {
                        randomPid = Math.floor(Math.random() * result.posts.$.count);
                    }
                    this._getImages(1, randomPid, tags).then(data => {
                        let clean = data.text.replace("\ufeff", "");
                        parseString(clean, (err, result) => {
                            if (err) reject(err);
                            if(this.safe === true){
                                if (testArraySfw(result)) {
                                    var image_url = result.posts.post[0].$.file_url;
                                    var tags = result.posts.post[0].$.tags;
                                    var source = result.posts.post[0].$.source;
                                    resolve({
                                        image_url: image_url,
                                        tags: tags,
                                        source: source,
                                        source_url: `https://safebooru.org/index.php?page=post&s=view&id=${result.posts.post[0].$.id}`
                                    });
                                } else {
                                    reject({
                                        error: "Missing ID or File URL"
                                    })
                                }
                            } else {
                                if (testArrayNsfw(result)) {
                                    var image_url = result.posts.post[0].file_url;
                                    var tags = result.posts.post[0].tags;
                                    var source = result.posts.post[0].source;
                                    resolve({
                                        image_url: image_url,
                                        tags: tags,
                                        source: source,
                                        source_url: `https://safebooru.org/index.php?page=post&s=view&id=${result.posts.post[0].id}`
                                    });
                                } else {
                                    reject({
                                        error: "Missing ID or File URL"
                                    })
                                }
                            }
                        })
                    })
                })
            })

        })
    }

    _getImages(limit, pid, tags) {
        return new Promise((resolve, reject) => {
            superagent.post(`${this.url}?page=dapi&s=post&q=index&limit=${limit}&pid=${pid}&tags=${tags}`)
                .end((err, res) => {
                    if (err) reject(err);
                    resolve(res);
                });
        });

    }
}

module.exports = {
    BooruGrabber
}
